# 01. Importing Libraries and Other Setup

In [7]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [8]:
# Create quick path to project folder

projpath = r'/Users/laineyodette/Documents/proDocs/professional development/Learning/Learning - Data Analytics/CareerFoundry - Become a Data Analyst/Data Immersion Course/A4 - Python Fundamentals for Data Analysts/Submitted Tasks/2024-01 Instacart Basket Analysis'

In [9]:
# Import products.csv dataset into df_prods df

df_prods = pd.read_csv(os.path.join(projpath, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [12]:
# Import products.csv dataset into df_ords df

df_ords = pd.read_csv(os.path.join(projpath, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = 0)

In [13]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 02. Data Consistency Checks Practice

In [14]:
# Create a test dataframe

df_test = pd.DataFrame()

In [15]:
# Create a mixed type column

df_test['mix'] = ['a','b', 1, True]

In [16]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [17]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [18]:
# Convert mix column to string data type

df_test['mix'] = df_test['mix'].astype('str')

# 03. Data Consistency Checks - df_prods

In [19]:
# Conduct initial descriptive analysis on prods df

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [20]:
# Find missing values by summing the number of null values in each column

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [21]:
# Create a new df to view null values

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [22]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [23]:
# Check to see how many rows are in the original dataframe

df_prods.shape

(49693, 5)

In [24]:
# Create a new dataframe without the rows with missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
df_prods_clean.shape

(49677, 5)

In [26]:
# Look for full duplicates in products df

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [27]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [28]:
df_prods_clean.shape

(49677, 5)

In [29]:
# Create new df removing duplicates 

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [30]:
df_prods_clean_no_dups.shape

(49672, 5)

In [31]:
# Export the checked products dataframe to 'Prepared Data' folder

df_prods_clean_no_dups.to_csv(os.path.join(projpath, '02 Data','Prepared Data', 'products_checked.csv'))

# 04. TASK Data Consistency Checks - df_ords

## Step 2 - Descriptive Analysis

In [32]:
# Conduct initial descriptive analysis on orders df

df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

**Observations**

- *Day of the week* is all under 7 (good since only 7 days in week and values should be between 0-6) - no outliers for min or max. 50th percentile is 3, which is correct.

- *Hour of the day* is all under 24 (good since only 24 hours in day and values should be between 0-23)

- *Days since prior order* should all be positive values, so these checks look okay.

## Step 3 - Mixed Data Types

In [33]:
# Check for mixed types

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

**Observations and Response**

- No mixed data type columns found in df_ords dataframe

- No fix needed (Step 4)

## Step 5 - Missing Values

In [34]:
# Find missing values by summing the number of null values in each column

df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

**Observations and Response**

- Missing values found in the "Days Since Prior Order" column

- This can be logical if a customer has not placed an order prior to the one in that row - in other words, this column would be null each time a new customer places their first order

## Step 6 - Fixing Missing Values

In [35]:
# Create boolean column for when the column days_since_prior_order is null to indicate it is the customer's first order

df_ords['customers_first_order'] = df_ords['days_since_prior_order'].isnull()*1

In [36]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  customers_first_order  
0                        8                     NaN                      1  
1                        7                    15.0                      0  
2                       12                    21.0                      0  
3                        7                    29.0                      0  
4                       15                    28.0                      0  
...                    ...                     ...                    ...  
3421078                 18                    29.0                      0  
3421079                 10                    30.0                      0  
3421080                 12                    18.0                      0  
3421081                 12                     7.0                      0  
3421082                 14                    30.0                      0  

[3421083 rows x 7 columns]

**Action Taken**

As I suspected that the 206,209 missing values corresponded to the first order of each customer, I looked at the descriptive stats above in the userID column. The maximum value was indeed 206209, meaning there are 206,209 customers, so my suspicions were true.

I decided to create a new column (a flag column) to indicate that the row with null values for "Days Since Prior Order" is the customer's first order (or not) - it indicates 1 or "true" if it's the customer's first order. It may be something interesting to call out in the report later to see how many subsequent orders were placed vs how many customers only placed a first order.

Note that I did not choose to impute, as the values missing are the story. There are no actual missing values. This is the same reason I did not remove the data - as the fact that the data is missing is the story itself.

## Step 7 - Duplicate Values

In [37]:
# Look for full duplicates in orders df

df_ords_dups = df_ords[df_ords.duplicated()]

In [38]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, customers_first_order]
Index: []

**Observations and Reactions**

- No duplicates were found in the orders df

- No action taken to address duplicates for Step 8

## Step 9 - Export Final Cleaned Data File

In [39]:
# Export the checked orders dataframe to 'Prepared Data' folder

df_ords.to_csv(os.path.join(projpath, '02 Data','Prepared Data', 'orders_checked.csv'))

**Note: df_prods file exported in Section 03 of code**